server/api/roster/roster-entity.ts

```diff
@@ -35,18 +53,135 @@ export class RosterEntity extends BaseEntity {
+      lastName: this.lastName,
+    };
+
+    for (const columnName of Object.keys(this.customColumns ?? {})) {
+      entryData[columnName] = this.customColumns[columnName];
+    }
+
+    return entryData;
+  }
+
   getColumnValue(column: RosterColumnInfo) {
     if (column.custom) {
-      return this.customColumns[column.name] || null;
+      return this.customColumns?.[column.name];
+    }
+
+    return Reflect.get(this, column.name);
+  }
+
+  setColumnValue(column: RosterColumnInfo, value: RosterColumnValue | undefined) {
+    // Ignore undefined values, unless this is a required column with no current value.
+    if (value === undefined) {
+      if (column.required && this.getColumnValue(column) != null) {
+        throw new RequiredColumnError(column.name);
+      }
+
+      return;
+    }
+
+    // Don't allow non-updatable columns to be updated if they already have a value.
+    if (!column.updatable && this.getColumnValue(column) !== undefined) {
+      throw new Error(`Column '${column.name}' is not updatable.`);
+    }
+
+    // If this is a null or empty value, make sure it's not a required column.
+    if (value === null || (typeof value === 'string' && value.length === 0)) {
+      if (column.required) {
+        throw new RequiredColumnError(column.name);
+      }
+    }
+
+    // Convert string data into the type our model expects.
+    if (typeof value === 'string' && column.type !== RosterColumnType.String) {
+      switch (column.type) {
+        case RosterColumnType.Number:
+          value = +value;
+          if (Number.isNaN(value)) {
+            throw new BadRequestError(`Number value for column '${column.name}' is invalid.`);
+          }
+          break;
+        case RosterColumnType.Date:
+        case RosterColumnType.DateTime:
+          value = dateFromString(value, true)!.toISOString();
+          break;
+        case RosterColumnType.Boolean:
+          value = (value === 'true');
+          break;
+        default:
+          break;
+      }
+    }
+
+    // Set the value.
+    if (column.custom) {
+      if (!this.customColumns) {
+        this.customColumns = {};
+      }
+      this.customColumns[column.name] = value;
+    } else {
+      Reflect.set(this, column.name, value);
     }
-    return Reflect.get(this, column.name) || null;
   }
+
+  setColumnValueFromData(column: RosterColumnInfo, data: RosterEntryData) {
+    // Get dates and enums as strings.
+    let paramType: BaseType;
+    switch (column.type) {
+      case RosterColumnType.Date:
+      case RosterColumnType.DateTime:
+      case RosterColumnType.Enum:
+        paramType = 'string';
+        break;
+      default:
+        paramType = column.type;
+    }
+
+    // Get the column value from the data.
+    let value: RosterColumnValue | undefined;
+    if (column.required) {
+      value = getRequiredValue(column.name, data, paramType);
+    } else {
+      value = getOptionalValue(column.name, data, paramType);
+    }
```

Comment:
I like the idea, and gave it a try but ended up aborting because it felt like it was obfuscating things too much. I'd rather have a little bit of repeated code and keep it from getting too abstract and murky. The fact that the file row version calls those functions in a different way is what complicates it.